TestNoDebug_AcceptNoRequestsButDisconnect fails intermittently

```
------- Stdout: -------
=== RUN   TestNoDebug_AcceptNoRequestsButDisconnect
DAP server listening at: [::]:50898
2022-01-04T15:24:01Z warning layer=rpc Listening for remote connections (connections are not authenticated nor encrypted)
2022-01-04T15:24:01Z debug layer=dap DAP server pid = 3868
Connecting to server at: [::]:50898
2022-01-04T15:24:01Z debug layer=dap DAP connection 70 started
2022-01-04T15:24:01Z debug layer=dap [<- from client]{"seq":1,"type":"request","command":"initialize","arguments":{"adapterID":"go","locale":"en-us","linesStartAt1":true,"columnsStartAt1":true,"pathFormat":"path","supportsVariableType":true,"supportsVariablePaging":true,"supportsRunInTerminalRequest":true}}
2022-01-04T15:24:01Z debug layer=dap [-> to client]{"seq":0,"type":"response","request_seq":1,"success":true,"command":"initialize","body":{"supportsConfigurationDoneRequest":true,"supportsFunctionBreakpoints":true,"supportsConditionalBreakpoints":true,"supportsEvaluateForHovers":true,"supportsSetVariable":true,"supportsExceptionInfoRequest":true,"supportTerminateDebuggee":true,"supportsDelayedStackTraceLoading":true,"supportsLogPoints":true,"supportsDisassembleRequest":true,"supportsClipboardContext":true,"supportsSteppingGranularity":true,"supportsInstructionBreakpoints":true}}
2022-01-04T15:24:01Z debug layer=dap [<- from client]{"seq":2,"type":"request","command":"launch","arguments":{"mode":"exec","noDebug":true,"program":"Z:\\buildAgent\\temp\\buildTmp\\http_server.b2c11601"}}
2022-01-04T15:24:01Z debug layer=dap parsed launch config: {
	"mode": "exec",
	"program": "Z:\\buildAgent\\temp\\buildTmp\\http_server.b2c11601",
	"noDebug": true,
	"backend": "default",
	"stackTraceDepth": 50
}
2022-01-04T15:24:01Z debug layer=dap launching binary 'Z:\buildAgent\temp\buildTmp\http_server.b2c11601' with config: {
	"mode": "exec",
	"program": "Z:\\buildAgent\\temp\\buildTmp\\http_server.b2c11601",
	"cwd": "Z:\\buildAgent\\work\\3aca8f960d346a9c\\service\\dap",
	"noDebug": true,
	"backend": "default",
	"stackTraceDepth": 50
}
2022-01-04T15:24:01Z debug layer=dap [-> to client]{"seq":0,"type":"response","request_seq":2,"success":true,"command":"launch"}
2022-01-04T15:24:01Z debug layer=dap [<- from client]{"seq":3,"type":"request","command":"setBreakpoints","arguments":{"source":{"name":"http_server.go","path":"Z:/buildAgent/work/3aca8f960d346a9c/_fixtures/http_server.go"},"breakpoints":[{"line":8}]}}
2022-01-04T15:24:01Z debug layer=dap noDebug mode: unable to process 'setBreakpoints' request
2022-01-04T15:24:01Z debug layer=dap [-> to client]{"seq":0,"type":"response","request_seq":3,"success":false,"command":"setBreakpoints","message":"noDebug mode","body":{"error":{"id":3000,"format":"noDebug mode: unable to process 'setBreakpoints' request","showUser":false}}}
2022-01-04T15:24:01Z debug layer=dap [<- from client]{"seq":4,"type":"request","command":"setFunctionBreakpoints","arguments":{"breakpoints":null}}
2022-01-04T15:24:01Z debug layer=dap noDebug mode: unable to process 'setFunctionBreakpoints' request
2022-01-04T15:24:01Z debug layer=dap [-> to client]{"seq":0,"type":"response","request_seq":4,"success":false,"command":"setFunctionBreakpoints","message":"noDebug mode","body":{"error":{"id":3000,"format":"noDebug mode: unable to process 'setFunctionBreakpoints' request","showUser":false}}}
2022-01-04T15:24:01Z debug layer=dap [<- from client]{"seq":5,"type":"request","command":"pause","arguments":{"threadId":1}}
2022-01-04T15:24:01Z debug layer=dap noDebug mode: unable to process 'pause' request
2022-01-04T15:24:01Z debug layer=dap [-> to client]{"seq":0,"type":"response","request_seq":5,"success":false,"command":"pause","message":"noDebug mode","body":{"error":{"id":3000,"format":"noDebug mode: unable to process 'pause' request","showUser":false}}}
2022-01-04T15:24:01Z debug layer=dap [<- from client]{"seq":6,"type":"request","command":"restart","arguments":{}}
2022-01-04T15:24:01Z debug layer=dap Unsupported command: cannot process "restart" request
2022-01-04T15:24:01Z debug layer=dap [-> to client]{"seq":0,"type":"response","request_seq":6,"success":false,"command":"restart","message":"Unsupported command","body":{"error":{"id":9999,"format":"Unsupported command: cannot process \"restart\" request","showUser":false}}}
2022-01-04T15:24:01Z debug layer=dap [<- from client]{"seq":7,"type":"request","command":"disconnect","arguments":{"terminateDebuggee":true}}
2022-01-04T15:24:01Z debug layer=dap [-> to client]{"seq":0,"type":"event","event":"output","body":{"category":"console","output":"Terminating process 2700\n","source":{}}}
2022-01-04T15:24:01Z debug layer=dap program exited with error: exit status 1
2022-01-04T15:24:01Z debug layer=dap [-> to client]{"seq":0,"type":"event","event":"output","body":{"category":"console","output":"Process 2700 has exited with status 1\n","source":{}}}
2022-01-04T15:24:01Z debug layer=dap noDebug process killed
2022-01-04T15:24:01Z debug layer=dap [-> to client]{"seq":0,"type":"response","request_seq":7,"success":true,"command":"disconnect"}
2022-01-04T15:24:01Z debug layer=dap [-> to client]{"seq":0,"type":"event","event":"terminated","body":{}}
2022-01-04T15:24:01Z debug layer=dap [-> to client]{"seq":0,"type":"event","event":"terminated","body":{}}
2022-01-04T15:24:01Z debug layer=dap DAP server stopping...
    server_test.go:5348: got &dap.DisconnectResponse{Response:dap.Response{ProtocolMessage:dap.ProtocolMessage{Seq:0, Type:"response"}, RequestSeq:7, Success:true, Command:"disconnect", Message:""}}, want *dap.TerminatedEvent
2022-01-04T15:24:01Z debug layer=dap DAP server stopped
--- FAIL: TestNoDebug_AcceptNoRequestsButDisconnect (6.24s)
```

After merging the other TestNoDebug fix.

cc @suzmue @polinasok

--- a/service/dap/daptest/client.go
+++ b/service/dap/daptest/client.go
@@ -150,7 +150,8 @@ const ProcessExited = `Process [0-9]+ has exited with status %s\n`
 
 func (c *Client) ExpectOutputEventProcessExited(t *testing.T, status int) *dap.OutputEvent {
 	t.Helper()
-	return c.ExpectOutputEventRegex(t, fmt.Sprintf(ProcessExited, fmt.Sprintf("%d", status)))
+	// We sometimes fail to return the correct exit status on Linux, so allow -1 here as well.
+	return c.ExpectOutputEventRegex(t, fmt.Sprintf(ProcessExited, fmt.Sprintf("(%d|-1)", status)))
 }
 
 func (c *Client) ExpectOutputEventProcessExitedAnyStatus(t *testing.T) *dap.OutputEvent {

--- a/service/dap/server_test.go
+++ b/service/dap/server_test.go
@@ -5353,11 +5353,39 @@ func TestNoDebug_AcceptNoRequestsButDisconnect(t *testing.T) {
 
 		// Disconnect request is ok
 		client.DisconnectRequestWithKillOption(true)
-		client.ExpectOutputEventTerminating(t)
-		client.ExpectOutputEventRegex(t, fmt.Sprintf(daptest.ProcessExited, "(-1|1)"))
-		client.ExpectTerminatedEvent(t)
-		client.ExpectDisconnectResponse(t)
-		client.ExpectTerminatedEvent(t)
+		terminated, disconnectResp := false, false
+		for {
+			m, err := client.ReadMessage()
+			if err != nil {
+				break
+			}
+			switch m := m.(type) {
+			case *dap.OutputEvent:
+				ok := false
+				wants := []string{`Terminating process [0-9]+\n`, fmt.Sprintf(daptest.ProcessExited, "(-1|1)")}
+				for _, want := range wants {
+					if matched, _ := regexp.MatchString(want, m.Body.Output); matched {
+						ok = true
+						break
+					}
+				}
+				if !ok {
+					t.Errorf("\ngot %#v\nwant Output=%q\n", m, wants)
+				}
+			case *dap.TerminatedEvent:
+				terminated = true
+			case *dap.DisconnectResponse:
+				disconnectResp = true
+			default:
+				t.Errorf("got unexpected message %#v", m)
+			}
+		}
+		if !terminated {
+			t.Errorf("did not get TerminatedEvent")
+		}
+		if !disconnectResp {
+			t.Errorf("did not get DisconnectResponse")
+		}
 	})
 }
 

